tests/test_util.py

```diff
@@ -30,3 +35,81 @@ def test_replace_control_chars(uncommon_text: str, sanitized_text: str) -> None:
     assert util.replace_control_chars(uncommon_text) == sanitized_text
     assert util.replace_control_chars("normal text") == "normal text"
     assert util.replace_control_chars("") == ""
+
+
+def test_nonblocking_read(mocker: MockerFixture) -> None:
+    """Test that the nonblocking_read() function works properly."""
+    size = 9
+    timeout = 1
+    r, w = os.pipe()
+
+    # Test 1 - Check that invalid arguments (blocking fd, negative size/timeout ) raise
+    # an exception.
+    with pytest.raises(ValueError, match="Expected a non-blocking file descriptor"):
+        util.nonblocking_read(r, size, timeout)
+
+    os.set_blocking(r, False)
+
+    with pytest.raises(ValueError, match="Expected a positive size value"):
+        util.nonblocking_read(r, 0, timeout)
+
+    with pytest.raises(ValueError, match="Expected a positive timeout value"):
+        util.nonblocking_read(r, size, 0)
+
+    # Test 2 - Check that partial reads are retried, for the timeout's duration,
+    # and we never read more than we want.
+    select_spy = mocker.spy(selectors.DefaultSelector, "select")
+    read_spy = mocker.spy(os, "read")
+
+    # Write "1234567890", with a delay of 0.3 seconds.
+    def write_rest() -> None:
+        os.write(w, b"12345")
```

Comment:
Sigh, this whole thing is racy. I can move `os.write(w, b"12345")` outside this function, so that the first select loop in `nonblocking_read` is sure to read it.

Reply:
See the fix in ce3c03349f115eab2ce6de2e9bc8e0806f039612.